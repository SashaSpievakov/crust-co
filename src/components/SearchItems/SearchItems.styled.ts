import styled from 'styled-components';
import { BsXLg, BsSearch } from 'react-icons/bs';

const Section = styled.section`
  position: relative;
`;

const SearchIcon = styled(BsSearch)`
  position: absolute;
  top: 10px;
  left: 10px;
  font-size: 20px;
  opacity: 0.5;

  &:hover {
    opacity: 0.7;
    cursor: pointer;
  }
`;

const Input = styled.input`
  font-size: 16px;
  color: ${({ theme }) => theme.colors.third};
  height: 40px;
  width: 240px;
  padding: 10px 33px 10px 40px;
  background-color: ${({ theme }) => theme.colors.gray};
  border: 1px solid rgba(0, 0, 0, 0.1);
  border-radius: 10px;
  transition: 0.3s;

  &:focus {
    width: 300px;
    border: 1px solid rgba(0, 0, 0, 0.3);
  }

  &::placeholder {
    color: ${({ theme }) => theme.colors.third};
    opacity: 0.6;
  }

  @media screen and (max-width: 800px) {
    width: 210px;

    &:focus {
      width: 250px;
    }
  }

  @media screen and (max-width: 400px) {
    width: 180px;

    &:focus {
      width: 240px;
    }
  }
`;

const Cross = styled(BsXLg)`
  position: absolute;
  top: 10px;
  right: 10px;
  font-size: 20px;
  opacity: 0.3;
  cursor: pointer;

  &:hover {
    opacity: 0.7;
  }
`;

export { Section, SearchIcon, Input, Cross };
