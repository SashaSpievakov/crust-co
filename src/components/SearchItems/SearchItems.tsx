import React, { memo, useCallback, useRef, useState } from 'react';
import debounce from 'lodash.debounce';

import { useAppDispatch, useAppSelector } from '../../hooks/reduxHooks';
import {
  selectSearchValue,
  setSearchValue,
} from '../../store/slices/searchSlice';
import { Wrapper, SearchIcon, Input, Cross } from './SearchItems.styled';

const SearchItems = memo(() => {
  const [value, setValue] = useState<string>('');
  const searchValue = useAppSelector(selectSearchValue);
  const dispatch = useAppDispatch();
  const inputRef = useRef<HTMLInputElement>(null);

  const updataSearchValue = useCallback(
    debounce((str: string) => {
      dispatch(setSearchValue(str));
    }, 200),
    [],
  );

  const onClickSearch = () => {
    inputRef.current?.focus();
  };

  const onChangeInput = (e: React.ChangeEvent<HTMLInputElement>): void => {
    setValue(e.target.value);
    updataSearchValue(e.target.value);
  };

  const onClickClose = () => {
    setValue('');
    dispatch(setSearchValue(''));
  };

  return (
    <Wrapper>
      <SearchIcon onClick={onClickSearch} data-testid="searchIcon" />
      <Input
        ref={inputRef}
        placeholder="Search..."
        value={value}
        onChange={(e) => onChangeInput(e)}
      />
      {searchValue && <Cross onClick={onClickClose} data-testid="cleanInput" />}
    </Wrapper>
  );
});

export default SearchItems;
