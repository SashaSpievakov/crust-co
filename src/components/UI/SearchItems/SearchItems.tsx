import { ChangeEventHandler, KeyboardEvent, memo, useRef } from 'react';

import { Cross, Input, SearchIcon, Wrapper } from './SearchItems.styled';

interface SearchItemsProps {
  value: string;
  width: string;
  onChange: ChangeEventHandler<HTMLInputElement>;
  onClickClose: () => void;
  expandedWidth?: string;
}

export const SearchItems = memo(
  ({
    value,
    width,
    onChange,
    onClickClose,
    expandedWidth,
  }: SearchItemsProps) => {
    const inputRef = useRef<HTMLInputElement | null>(null);

    const onClickSearch = () => {
      inputRef.current?.focus();
    };

    const handleCrossKeyDownClick = (e: KeyboardEvent<SVGElement>) => {
      if (e.code === 'Enter') {
        onClickClose();
      }
    };

    return (
      <Wrapper>
        <SearchIcon onClick={onClickSearch} data-testid="searchIcon" />
        <Input
          ref={inputRef}
          placeholder="Search..."
          value={value}
          onChange={onChange}
          width={width}
          expandedWidth={expandedWidth}
        />
        {value && (
          <Cross
            onClick={onClickClose}
            onKeyDown={(e) => handleCrossKeyDownClick(e)}
            tabIndex={0}
            data-testid="cleanInput"
          />
        )}
      </Wrapper>
    );
  },
);

SearchItems.displayName = 'SearchItems';
