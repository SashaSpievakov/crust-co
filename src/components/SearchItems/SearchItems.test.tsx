import '@testing-library/jest-dom';
import { screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import SearchItems from './SearchItems';
import rendererWithAllProviders from '../../tests/helpers/rendererWithProviders';
import renderWithAllProviders from '../../tests/helpers/renderWithProviders';

describe('SearchItems Test', () => {
  test('renders the SearchItems component', () => {
    const snapshot = rendererWithAllProviders(<SearchItems />);
    expect(snapshot).toMatchSnapshot();
  });

  describe('checks SearchItems functionality', () => {
    beforeEach(() => {
      renderWithAllProviders(<SearchItems />);
    });

    test('focuses on icon click', () => {
      const input = screen.getByRole('textbox');
      const searchIcon = screen.getByTestId('searchIcon');

      userEvent.click(searchIcon);

      expect(input).toHaveFocus();
    });

    test('types text into the input', () => {
      const input = screen.getByRole('textbox');

      userEvent.type(input, 'pizza');

      expect(input).toHaveDisplayValue('pizza');
    });

    test('shows delete text button', async () => {
      const input = screen.getByRole('textbox');

      userEvent.type(input, 'pizza');

      await waitFor(() => {
        expect(screen.getByTestId('cleanInput')).toBeInTheDocument();
      });
    });

    test('deletes text on click', async () => {
      const input = screen.getByRole('textbox');

      userEvent.type(input, 'pizza');

      await waitFor(() => {
        const deleteIcon = screen.getByTestId('cleanInput');
        userEvent.click(deleteIcon);
      });

      expect(input).toHaveDisplayValue('');
      expect(input).not.toHaveFocus();
    });
  });
});
