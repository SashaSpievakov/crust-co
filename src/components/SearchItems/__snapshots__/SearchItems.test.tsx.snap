// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchItems Test renders the SearchItems component 1`] = `
Object {
  "children": Array [
    <svg
      className="sc-gswNZR hNTceg"
      data-testid="searchIcon"
      fill="currentColor"
      height="1em"
      onClick={[Function]}
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 16 16"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M11.742 10.344a6.5 6.5 0 1 0-1.397 1.398h-.001c.03.04.062.078.098.115l3.85 3.85a1 1 0 0 0 1.415-1.414l-3.85-3.85a1.007 1.007 0 0 0-.115-.1zM12 6.5a5.5 5.5 0 1 1-11 0 5.5 5.5 0 0 1 11 0z"
      />
    </svg>,
    <input
      className="sc-dkrFOg jqvJmq"
      onChange={[Function]}
      placeholder="Search..."
      value=""
    />,
  ],
  "props": Object {
    "className": "sc-bcXHqe zqhWb",
  },
  "type": "div",
}
`;
